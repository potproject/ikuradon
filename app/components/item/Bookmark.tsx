import React, { useContext, useState, memo } from "react";
import { View, StyleSheet, TouchableOpacity } from "react-native";
import { FontAwesome } from "@expo/vector-icons";
import PropTypes from "prop-types";

import { ThemeContext } from "react-native-elements";

function Bookmark({ id, tootid, style, bookmarked, onBookmark }){
    const [stateBookmarked, useStateBookmarked] = useState(bookmarked);
    const { theme } = useContext(ThemeContext);
    return (
        <View style={[style, styles.container]}>
            <TouchableOpacity style={[style, { flex: 1 }]} onPress={() => {
                useStateBookmarked(!stateBookmarked);
                onBookmark(id, tootid, !stateBookmarked);
            }}>
                <FontAwesome name="bookmark" size={20} color={stateBookmarked ? theme.customColors.item.bookmark : theme.customColors.item.none} />
            </TouchableOpacity>
        </View>
    );
}

Bookmark.propTypes = {
    id: PropTypes.string,
    tootid: PropTypes.string,
    style: PropTypes.object,
    bookmarked: PropTypes.bool,
    count: PropTypes.number,
    onFavourite: PropTypes.func
};

const styles = StyleSheet.create({
    container: {
        flex: 1,
        flexDirection: "row"
    }
});

export default memo(Bookmark, (p, n) => p.id === n.id && p.favourited === n.favourited);