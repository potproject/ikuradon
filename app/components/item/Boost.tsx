import React, { useContext, useState, memo } from "react";
import { View, Text, StyleSheet, TouchableOpacity } from "react-native";
import { FontAwesome } from "@expo/vector-icons";
import PropTypes from "prop-types";

import { ThemeContext } from "react-native-elements";

function Boost(props){
    const { id, tootid, style, reblogged, count, disabled, onBoost } = props;
    const [stateReblogged, useStateReblogged] = useState(reblogged);
    const { theme } = useContext(ThemeContext);
    return (
        <View style={[style, styles.container]}>
            <TouchableOpacity style={style} disabled={disabled} onPress={() => {
                useStateReblogged(!stateReblogged);
                onBoost(id, tootid, !stateReblogged);
            }}>
                <FontAwesome name={disabled ? "lock" : "retweet"} size={20} color={stateReblogged === true ? theme.customColors.item.boost : theme.customColors.item.none} />
            </TouchableOpacity>
            <Text style={[{ color: theme.colors.grey0, fontSize: count > 99 ? 10 : 16 }, styles.text]}>{count !== 0 ? count : ""}</Text>
        </View>
    );
}

Boost.propTypes = {
    id: PropTypes.string,
    tootid: PropTypes.string,
    style: PropTypes.object,
    reblogged: PropTypes.bool,
    count: PropTypes.number,
    disabled: PropTypes.bool,
    onBoost: PropTypes.func
};

const styles = StyleSheet.create({
    container: {
        flex: 1,
        flexDirection: "row"
    },
    text: {
        flex: 1
    }
});

export default memo(Boost, (p, n) => p.id === n.id && p.reblogged === n.reblogged);