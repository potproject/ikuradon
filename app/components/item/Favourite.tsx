import React, { useContext, useState, memo } from "react";
import { View, Text, StyleSheet, TouchableOpacity } from "react-native";
import { FontAwesome } from "@expo/vector-icons";
import PropTypes from "prop-types";

import { ThemeContext } from "react-native-elements";

function Favourite({ id, tootid, style, favourited, count, onFavourite }){
    const [stateFavourited, useStateFavourited] = useState(favourited);
    const { theme } = useContext(ThemeContext);
    return (
        <View style={[style, styles.container]}>
            <TouchableOpacity style={[style, { flex: 1 }]} onPress={() => {
                useStateFavourited(!stateFavourited);
                onFavourite(id, tootid, !stateFavourited);
            }}>
                <FontAwesome name="star" size={20} color={stateFavourited ? theme.customColors.item.favourite : theme.customColors.item.none} />
            </TouchableOpacity>
            <Text style={[{ color: theme.colors.grey0, fontSize: count > 99 ? 10 : 16 }, styles.text]}>{count !== 0 ? count : ""}</Text>
        </View>
    );
}

Favourite.propTypes = {
    id: PropTypes.string,
    tootid: PropTypes.string,
    style: PropTypes.object,
    favourited: PropTypes.bool,
    count: PropTypes.number,
    onFavourite: PropTypes.func
};

const styles = StyleSheet.create({
    container: {
        flex: 1,
        flexDirection: "row"
    },
    text: {
        flex: 1
    }
});

export default memo(Favourite, (p, n) => p.id === n.id && p.favourited === n.favourited);