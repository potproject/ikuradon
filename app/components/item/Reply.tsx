import React, { useContext, memo } from "react";
import { TouchableOpacity, View, StyleSheet } from "react-native";
import { FontAwesome } from "@expo/vector-icons";
import PropTypes from "prop-types";

import { ThemeContext } from "react-native-elements";

function Reply({ id, tootid, user, acct, image, body, style, onReply }){
    const { theme } = useContext(ThemeContext);
    return (
        <View style={[style, styles.container]}>
            <TouchableOpacity style={styles.container} onPress={() => onReply(id, tootid, user, acct, image, body)}>
                <FontAwesome name="reply" size={20} color={theme.customColors.item.none} />
            </TouchableOpacity>
            <View style={styles.container} />
        </View>
    );
}

Reply.propTypes = {
    id: PropTypes.string,
    tootid: PropTypes.string,
    user: PropTypes.string,
    acct: PropTypes.string,
    image: PropTypes.string,
    body: PropTypes.string,
    style: PropTypes.object,
    onReply: PropTypes.func
};

const styles = StyleSheet.create({
    container: {
        flex: 1,
        flexDirection: "row"
    }
});

export default memo(Reply, (p, n) => p.id === n.id);