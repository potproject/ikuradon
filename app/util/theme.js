import defaultTheme from "../themes/default";
import MikuGreenTheme from "../themes/mikugreen";
import TootBlueTheme from "../themes/tootblue";
import DarkTheme from "../themes/dark";

export function settingTheme(updateTheme, value){
    switch (value){
        case "mikugreen":
            updateTheme(MikuGreenTheme);
            return;
        case "tootblue":
            updateTheme(TootBlueTheme);
            return;
        case "dark":
            updateTheme(DarkTheme);
            return;
        default:
            updateTheme(defaultTheme);
            return;
    }
}