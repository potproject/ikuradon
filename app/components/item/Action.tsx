import React, { memo, useContext } from "react";
import { TouchableOpacity, Clipboard, View, StyleSheet } from "react-native";
import { FontAwesome } from "@expo/vector-icons";
import { useActionSheet } from "@expo/react-native-action-sheet";
import t from "../../services/I18n";
import { open as openUrl } from "../../util/url";
import { bodyFormat, bodyExtractionUrl } from "../../util/parser";
import PropTypes from "prop-types";
import { ThemeContext } from "react-native-elements";

function Action({ id, tootid, style, url, account_url, user, acct, image, body, myself, onReply, onHide, onDeleting }){
    const { theme } = useContext(ThemeContext);
    const { showActionSheetWithOptions } = useActionSheet();
    const onOpenActionSheet = () => {
        let cancelButtonIndex = 6;
        let destructiveButtonIndex = 5;
        let options = [t("action_openinbrowser"), t("action_openinbrowserprofile"), t("action_copy"), t("action_copyurl"), t("action_reply"), t("action_hide")];
        // 自分のtootなら削除可能に
        if (myself) {
            options.push(t("action_delete"));
            cancelButtonIndex++;
            destructiveButtonIndex++;
        }
        options.push(t("global_cancel"));
        showActionSheetWithOptions(
            {
                options,
                cancelButtonIndex,
                destructiveButtonIndex,
            },
            buttonIndex => {
                switch (buttonIndex) {
                    case 0: //Open in Browser
                        openUrl(url);
                        return;
                    case 1: //Open in Browser Profile
                        openUrl(account_url);
                        return;
                    case 2: //Copy
                        Clipboard.setString(bodyFormat(body));
                        return;
                    case 3: //URL Copy
                        Clipboard.setString(bodyExtractionUrl(body));
                        return;
                    case 4: //reply
                        onReply(id, tootid, user, acct, image, body);
                        return;
                    case 5: //Hide
                        onHide(id);
                        return;
                    case 6: //Delete
                        if (myself) {
                            onDeleting(id);
                        }
                        return;
                }
            },
        );
    };
    return (
        <View style={[style, styles.container]}>
            <TouchableOpacity style={style} onPress={() => onOpenActionSheet()}>
                <FontAwesome name="ellipsis-h" size={20} color={theme.customColors.item.none} />
            </TouchableOpacity>
            <View style={styles.container} />
        </View>
    );
}

Action.propTypes = {
    id: PropTypes.string,
    tootid: PropTypes.string,
    style: PropTypes.object,
    url: PropTypes.string,
    account_url: PropTypes.string,
    user: PropTypes.string,
    acct: PropTypes.string,
    image: PropTypes.string,
    body: PropTypes.string,
    myself: PropTypes.bool,

    onReply: PropTypes.func,
    onHide: PropTypes.func,
    onDeleting: PropTypes.func
};

const styles = StyleSheet.create({
    container: {
        flex: 1,
        flexDirection: "row"
    }
});

export default memo(Action, (p, n) => p.id === n.id);