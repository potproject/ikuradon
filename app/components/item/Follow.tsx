import React, { useContext, useState, useEffect } from "react";
import { View, Text, StyleSheet, TouchableOpacity } from "react-native";
import { FontAwesome } from "@expo/vector-icons";
import PropTypes from "prop-types";

import { ThemeContext } from "react-native-elements";
import t from "../../services/I18n";
import { getRelationship } from "../../util/relationships";

function Follow({ id, style, onFollow }){
    const [stateFollowed, useStateFollowed] = useState(false);
    const { theme } = useContext(ThemeContext);
    useEffect(() => {
        getRelationship(id).then(({ data, error }) => {
            if (error === null && data.following){
                useStateFollowed(true);
            }
        });
    }, []);
    return (
        <View style={[style, styles.container]}>
            <TouchableOpacity style={style} onPress={() => {
                useStateFollowed(!stateFollowed);
                onFollow(id, !stateFollowed);
            }}>
                <Text style={styles.text}>
                    <FontAwesome name={stateFollowed ? "user" : "user-plus"} size={26} color={stateFollowed ? theme.colors.primary : theme.customColors.item.none} />
                    <Text style={[{ color: stateFollowed ? theme.colors.primary : theme.customColors.item.none }, styles.inlineText]}>
                        {" "}{stateFollowed ? t("notifications.unfollow") : t("notifications.follow")}
                    </Text>
                </Text>
            </TouchableOpacity>
        </View>
    );
}

Follow.propTypes = {
    id: PropTypes.string,
    userid: PropTypes.string,
    style: PropTypes.object,
    onFollow: PropTypes.func
};

const styles = StyleSheet.create({
    container: {
        flex: 1
    },
    text: {
        fontSize: 16,
        fontWeight: "bold",
    },
});

export default Follow;