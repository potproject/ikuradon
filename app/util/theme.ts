import defaultTheme from "../themes/default";
import MikuGreenTheme from "../themes/mikugreen";
import TootBlueTheme from "../themes/tootblue";
import DarkTheme from "../themes/dark";
import MikuGreenDarkTheme from "../themes/mikugreendark";
import TootBlueDarkTheme from "../themes/tootbluedark";

import * as Storage from "../util/storage";
import * as CONST_Storage from "../constants/storage";

export async function getTheme(){
    const config = await Storage.getItem(CONST_Storage.Config);
    let themeName = "default";
    if (config !== null && typeof config.theme !== "undefined"){
        themeName = config.theme;
    }
    switch (themeName){
        case "mikugreen":
            return MikuGreenTheme;
        case "tootblue":
            return TootBlueTheme;
        case "dark":
            return DarkTheme;
        case "mikugreendark":
            return MikuGreenDarkTheme;
        case "tootbluedark":
            return TootBlueDarkTheme;
        default:
            return defaultTheme;
    }
}