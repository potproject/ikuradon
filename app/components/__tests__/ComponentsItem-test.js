import React from "react";
import { render, fireEvent } from "@testing-library/react-native";

import theme from "../../themes/default";
import { ThemeContext } from "react-native-elements";

import Action from "../item/Action";
import Bookmark from "../item/Bookmark";
import Boost from "../item/Boost";
import Favourite from "../item/Favourite";
import Follow from "../item/Follow";
import Reply from "../item/Reply";

import ExampleStatus from "../../example/status";
import ExampleAccount from "../../example/account";

jest.mock("@expo/vector-icons/build/FontAwesome", () => "FontAwesome");
jest.mock("react-native-elements/src/icons/Icon", () => "Icon");

it("<Action />", () => {
    const result = render(
        <ThemeContext.Provider value={{ theme }}>
            <Action
                id={ExampleStatus().id}
                tootid={ExampleStatus().id}
                style={{}}
                url={ExampleStatus().url}
                account_url={ExampleAccount().url}
                user={ExampleAccount().display_name}
                acct={ExampleAccount().acct}
                image={ExampleAccount().avatar}
                body={ExampleAccount().content}
                myself={true}
                onReply={() => null}
                onHide={() => null}
                onDeleting={() => null}
            />
        </ThemeContext.Provider>
    );
    expect(result).toMatchSnapshot();
});

it("<Bookmark />", () => {
    const result = render(
        <ThemeContext.Provider value={{ theme }}>
            <Bookmark
                id={ExampleStatus().id}
                tootid={ExampleStatus().id}
                bookmarked={true}
                style={{}}
                onBookmark={() => null}
            />
        </ThemeContext.Provider>
    );
    expect(result).toMatchSnapshot();
});

it("<Boost />", () => {
    const result = render(
        <ThemeContext.Provider value={{ theme }}>
            <Boost
                id={ExampleStatus().id}
                tootid={ExampleStatus().id}
                reblogged={true}
                count={1}
                style={{}}
                onBoost={() => null}
                disabled={false}
            />
        </ThemeContext.Provider>
    );
    expect(result).toMatchSnapshot();
});

it("<Favourite />", () => {
    const result = render(
        <ThemeContext.Provider value={{ theme }}>
            <Favourite
                id={ExampleStatus().id}
                tootid={ExampleStatus().id}
                favourited={true}
                count={2}
                style={{}}
                onFavourite={()=>null}
            />
        </ThemeContext.Provider>
    );
    expect(result).toMatchSnapshot();
});

it("<Follow />", () => {
    const result = render(
        <ThemeContext.Provider value={{ theme }}>
            <Follow
                id={ExampleStatus().id}
                style={{}}
                onFollow={() => null}
            />
        </ThemeContext.Provider>
    );
    expect(result).toMatchSnapshot();
});

it("<Reply />", () => {
    const result = render(
        <ThemeContext.Provider value={{ theme }}>
            <Reply
                id={ExampleStatus().id}
                tootid={ExampleStatus().id}
                user={ExampleAccount().display_name}
                acct={ExampleAccount().acct}
                image={ExampleAccount().avatar}
                body={ExampleStatus().content}
                style={{}}
                onReply={() => null}
            />
        </ThemeContext.Provider>
    );
    expect(result).toMatchSnapshot();
});