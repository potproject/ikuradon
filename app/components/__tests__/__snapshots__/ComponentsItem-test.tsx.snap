// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Action /> 1`] = `
<View
  style={
    Array [
      Object {},
      Object {
        "flex": 1,
        "flexDirection": "row",
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <FontAwesome
      color="#8899a6"
      name="ellipsis-h"
      size={20}
    />
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
      }
    }
  />
</View>
`;

exports[`<Bookmark /> 1`] = `
<View
  style={
    Array [
      Object {},
      Object {
        "flex": 1,
        "flexDirection": "row",
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 1,
        "opacity": 1,
      }
    }
  >
    <FontAwesome
      color="#4c6ef5"
      name="bookmark"
      size={20}
    />
  </View>
</View>
`;

exports[`<Boost /> 1`] = `
<View
  style={
    Array [
      Object {},
      Object {
        "flex": 1,
        "flexDirection": "row",
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <FontAwesome
      color="#2b90d9"
      name="retweet"
      size={20}
    />
  </View>
  <Text
    style={
      Array [
        Object {
          "color": "#657786",
          "fontSize": 16,
        },
        Object {
          "flex": 1,
        },
      ]
    }
  >
    1
  </Text>
</View>
`;

exports[`<Favourite /> 1`] = `
<View
  style={
    Array [
      Object {},
      Object {
        "flex": 1,
        "flexDirection": "row",
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 1,
        "opacity": 1,
      }
    }
  >
    <FontAwesome
      color="#ffd27d"
      name="star"
      size={20}
    />
  </View>
  <Text
    style={
      Array [
        Object {
          "color": "#657786",
          "fontSize": 16,
        },
        Object {
          "flex": 1,
        },
      ]
    }
  >
    2
  </Text>
</View>
`;

exports[`<Follow /> 1`] = `
<View
  style={
    Array [
      Object {},
      Object {
        "flex": 1,
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      <FontAwesome
        color="#8899a6"
        name="user-plus"
        size={26}
      />
      <Text
        style={
          Array [
            Object {
              "color": "#8899a6",
            },
            undefined,
          ]
        }
      >
         
        Follow
      </Text>
    </Text>
  </View>
</View>
`;

exports[`<Reply /> 1`] = `
<View
  style={
    Array [
      Object {},
      Object {
        "flex": 1,
        "flexDirection": "row",
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "opacity": 1,
      }
    }
  >
    <FontAwesome
      color="#8899a6"
      name="reply"
      size={20}
    />
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
      }
    }
  />
</View>
`;
